### PR TITLE
Feature/activate rewards

### PR DESCRIPTION
**Pull request by iStrike7**

### CODE REVIEW

File: pallets/xyk/src/lib.rs [hunk 1]
```diff
@@ -864,6 +915,29 @@ impl<T: Config> Pallet<T> {
...
 		);
 
+		LiquidityMiningActiveUser::<T>::try_mutate((&user, liquidity_asset_id), |active_amount| {
+			if let Some(val) = active_amount.checked_add(liquidity_assets_added) {
+				*active_amount = val;
+				Ok(())
+			} else {
+				Err(())
+			}
+		});
```
Comment (iStrike7) [inline pallets/xyk/src/lib.rs:925]: If this fails, the rest will continue to execute, right?
That doesn't seem safe.

File: pallets/xyk/src/lib.rs [hunk 2]
```diff
@@ -595,6 +607,36 @@ pub mod pallet {
...
+			amount: Balance,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			<Self as XykFunctionsTrait<T::AccountId>>::activate_liquidity(
+				sender,
+				liquidity_token_id,
+				amount,
+			)
+		}
```
Comment (iStrike7) [inline pallets/xyk/src/lib.rs:624]: I suggest you use the transactional annotation, like Mat has used in the bootstrap pallet.
You should use it for all extrinsics, EXCEPT for buy and sell asset as they need to charge percent for every txn even if it fails.

File: pallets/xyk/src/lib.rs [hunk 3]
```diff
@@ -2074,28 +2186,17 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 		let liquidity_asset_id = Pallet::<T>::get_liquidity_asset(first_asset_id, second_asset_id)?;
 
 		// Ensure user has enought liquidity tokens to burn
+		let liquidity_token_free_balance =
+			<T as Config>::Currency::free_balance(liquidity_asset_id.into(), &sender).into();
+		let liquidity_token_activated_balance =
+			LiquidityMiningActiveUser::<T>::get((&sender, &liquidity_asset_id));
+		let total_liquidity_tokens_user_owned =
+			liquidity_token_free_balance + liquidity_token_activated_balance;
```
Comment (iStrike7) [inline pallets/xyk/src/lib.rs:2194]: While this does seem safe, I would strongly recommend you to use checked math everywhere as this is a complicated pallet. and even with staking which was forked from moonbeam we had issues with unsafe math.
You needn't worry much about ensuring that all checks will pass before making changes to storage, if you use the transactional annotation.

Comment (iStrike7) [reply pallets/xyk/src/lib.rs:2194]: But make sure to keep in mind the exact flow when relying on it, as there may be some shared functions that show up in both transactional and non-transactional callers.

**Review by iStrike7 — COMMENTED** [review]
@goncer Had come across some bugs as well. Please look into that too.

I have not reviewed the logic, or the code flow.

File: pallets/xyk/src/lib.rs [hunk 4]
```diff
@@ -2307,29 +2426,63 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
...
+		ensure!(
+			<T as Config>::PoolPromoteApi::get_pool_rewards(liquidity_asset_id).is_some(),
+			Error::<T>::NotAPromotedPool
 		);
-		Pallet::<T>::deposit_event(Event::PoolPromoted(liquidity_token_id));
+		ensure!(
+			<T as Config>::Currency::free_balance(liquidity_asset_id.into(), &user).into() >=
+				amount,
+			Error::<T>::NotEnoughAssets
+		);
```
Comment (iStrike7) [inline pallets/xyk/src/lib.rs:2460]: This should be ensure_can_withdraw.
Frozen/locked tokens are not accounted for in free_balance.

Comment (iStrike7) [reply pallets/xyk/src/lib.rs:2460]: Please make similar changes across the pallet.
This is a major breaking issue

Comment (iStrike7) [reply pallets/xyk/src/lib.rs:2460]: Also you mentioned "somewhere I need “disposable” amount". Please make sure that this value is computed as free.saturating_sub(frozen)

File: pallets/xyk/src/lib.rs (hunk 1 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs]: added checked error:

`LiquidityMiningActivePool::<T>::try_mutate(liquidity_asset_id, |active_amount| {
			if let Some(val) = active_amount.checked_add(liquidity_assets_added) {
				*active_amount = val;
				Ok(())
			} else {
				Err(())
			}
		})
		.map_err(|_| DispatchError::from(Error::<T>::MathOverflow))?;`

File: pallets/xyk/src/tests.rs [hunk 5]
```diff
@@ -1314,6 +1315,174 @@ fn liquidity_rewards_deactivate_W() {
 	});
 }
 
+#[test]
+#[serial]
+fn liquidity_rewards_calculate_work_after_deactivate() {
```
Comment (mateuszaaa) [inline pallets/xyk/src/tests.rs:1320]: is that test related to the impl from this pr ?

**Review by mateuszaaa — CHANGES_REQUESTED** [review]
few smaller comments

File: pallets/xyk/src/lib.rs (hunk 4 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs]: added free-frozen as disposable amount

File: pallets/xyk/src/tests.rs [hunk 6]
```diff
@@ -509,6 +509,7 @@ fn liquidity_rewards_claim_more_NW() {
 fn liquidity_rewards_burn_W() {
 	new_test_ext().execute_with(|| {
 		initialize_liquidity_rewards();
+		//	env_logger::init();
```
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/xyk/src/tests.rs]: removed

File: pallets/xyk/src/lib.rs [hunk 7]
```diff
@@ -2245,14 +2248,21 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 		// Ensure user has enought liquidity tokens to burn
 		let liquidity_token_free_balance =
 			<T as Config>::Currency::free_balance(liquidity_asset_id.into(), &sender).into();
+		let liquidity_token_locked_balance =
```
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:2249]: warnings addressed

[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:2249]: warnings addressed

File: pallets/xyk/src/tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/xyk/src/tests.rs:1304]: tests are related to some scenarios which were addressed in this and previous pr

Comment (stanieltron) [reply pallets/xyk/src/tests.rs:1304]: tests are related to some scenarios which were addressed in this and previous pr

**Review by mateuszaaa — APPROVED** [review]
LGTM
